fix(json-type): add support for JSON columns in the service dashboard

Perseus opened the pull request:
Fixes #1023 

- Adds a `useJsonColumns` hook which returns the names of all JSON columns
- Modifies the syntax used to access attributes in the traces table based on the type of the column
- Changes some `coalesce(nullIf(...))` queries to access the fields directly due to ClickHouse not supporting `NULLIF()` on dynamic columns

knudtty commented on packages/app/src/serviceDashboard.ts:
```diff
@@ -1,7 +1,37 @@
+  const isJsonColumn = jsonColumns.includes(spanAttributeField);
+
+  const formatFieldAccess = (field: string, key: string) => {
+    if (isJsonColumn) {
+      return `${field}.\`${key}\``;
+    } else {
+      return `${field}['${key}']`;
+    }
```
Could this be a ternary?

knudtty commented on packages/app/src/serviceDashboard.ts:
```diff
@@ -1,7 +1,37 @@
+  // existence of the serverAddress/httpHost to build the span name
+  if (isJsonColumn) {
+    dbStatement = `
+      coalesce(
+        if(
+          toString(${formatFieldAccess(spanAttributeField, 'db.query.text')}) != '',
+          toString(${formatFieldAccess(spanAttributeField, 'db.query.text')}),
+          if(
+            toString(${formatFieldAccess(spanAttributeField, 'db.statement')}) != '',
+            toString(${formatFieldAccess(spanAttributeField, 'db.statement')}), 
+            ''
+          )
+        )
+      )
```
I'm noticing this is similar to the `spanNameColSql`, is there any way to generalize/extract this? "No" is a fine answer btw

wrn14897 commented on packages/common-utils/src/metadata.ts:
```diff
@@ -147,6 +147,27 @@ export class Metadata {
+    });
+
+    // TODO: should we use .includes() to handle Array(JSON) and other variants?
+    return columns
+      .filter(column => column.type.startsWith('JSON'))
+      .map(column => column.name);
```
style: can use `filterColumnMetaByType` https://github.com/hyperdxio/hyperdx/blob/86115fa58a537248557731617baf6830238da30c/packages/common-utils/src/clickhouse.ts#L741

wrn14897 commented on packages/app/src/hooks/useMetadata.tsx:
```diff
@@ -43,6 +43,33 @@ export function useColumns(
+    queryKey: ['useMetadata.useJsonColumns', { databaseName, tableName }],
+    queryFn: async () => {
+      const metadata = getMetadata();
+      return metadata.getJsonColumns({
```
style: we can reuse `getColumns` and do the filtering afterward unless `getJsonColumns` is going to be used somewhere else

wrn14897 commented on packages/app/src/serviceDashboard.ts:
```diff
@@ -1,7 +1,37 @@
+    }
+  };
+
+  let dbStatement = `coalesce(nullif(${formatFieldAccess(spanAttributeField, 'db.query.text')}, ''), nullif(${formatFieldAccess(spanAttributeField, 'db.statement')}, ''))`;
```
Great catch on the outdated semantic convention! we can create a generic coalesce select generator so it can be reused for future semantic convention migration especially the custom dashboards

Perseus commented:
@wrn14897 @knudtty
Thanks for the review!

I've made a function to generalize the creation of coalesced queries based on both of your comments; it feels a little awkward since its building the queries manually and I've tried to keep the queries well-indented to make troubleshooting easier without adding significant amounts of whitespace, but it would be significantly easier if there was a standardized query builder (or I could also write a query builder specifically for this use-case if that sounds better).

I've addressed the remaining comments as well.

knudtty reviewed (approved):
This is a great contribution! Thank you for the fix
